proxy/src/redis/kv_ops.rs

```diff
@@ -0,0 +1,305 @@
+                        _guard,
+                    } => {
+                        if let Some(resp_tx) = resp_tx {
+                            drop(resp_tx.send(self.hset(key, field, value).await));
+                        } else {
+                            drop(self.hset(key, field, value).await);
+                        }
+                    }
+                    RedisOp::HSetMultiple {
+                        key,
+                        items,
+                        resp_tx,
+                        _guard,
+                    } => {
+                        if let Some(resp_tx) = resp_tx {
+                            drop(resp_tx.send(self.hset_multiple(&key, &items).await));
+                        } else {
+                            drop(self.hset_multiple(&key, &items).await);
+                        }
+                    }
+                    RedisOp::Expire {
+                        key,
+                        seconds,
+                        resp_tx,
+                        _guard,
+                    } => {
+                        if let Some(resp_tx) = resp_tx {
+                            drop(resp_tx.send(self.expire(&key, seconds).await));
+                        } else {
+                            drop(self.expire(&key, seconds).await);
+                        }
+                    }
+                    RedisOp::HGet {
+                        key,
+                        field,
+                        resp_tx,
+                        _guard,
+                    } => {
+                        drop(resp_tx.send(self.hget(key, field).await));
+                    }
+                    RedisOp::HGetAll {
+                        key,
+                        resp_tx,
+                        _guard,
+                    } => {
+                        drop(resp_tx.send(self.hget_all(key).await));
+                    }
+                    RedisOp::HDel {
+                        key,
+                        field,
+                        resp_tx,
+                        _guard,
+                    } => {
+                        if let Some(resp_tx) = resp_tx {
+                            drop(resp_tx.send(self.hdel(key, field).await));
+                        } else {
+                            drop(self.hdel(key, field).await);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn hset<K, F, V>(&mut self, key: K, field: F, value: V) -> anyhow::Result<()>
+    where
+        K: ToRedisArgs + Send + Sync + Clone,
+        F: ToRedisArgs + Send + Sync + Clone,
+        V: ToRedisArgs + Send + Sync + Clone,
+    {
+        if !self.limiter.check() {
+            tracing::info!("Rate limit exceeded. Skipping hset");
+            return Err(anyhow::anyhow!("Rate limit exceeded"));
+        }
+
+        match self
+            .client
+            .hset(key.clone(), field.clone(), value.clone())
```

Comment:
`&impl ToRedisArgs` also implements `ToRedisArgs`, so instead of cloning you should be able to pass references here.

Reply:
I would like to do so, but we need cloning for retry later in this function:

```
        tracing::info!("Redis client is disconnected. Reconnectiong...");
        self.try_connect().await?;
        self.client
            .hset(key, field, value)
            .await
            .map_err(anyhow::Error::new)

```